### PR TITLE
Refactor distr_test.go to table-driven

### PR DESCRIPTION
**Pull request by hieuvubk**
##  I just refactor some tests in distr_test.go to table-driven for issue #1961:
 
  - *TestAllocateAsset*
  - *TestReplaceDistrRecords*
  - *TestUpdateDistrRecords*

### CODE REVIEW

File: x/pool-incentives/keeper/distr.go
Comment (ValarDragon) [inline x/pool-incentives/keeper/distr.go:45]: should be deleted right?

Comment (hieuvubk) [reply x/pool-incentives/keeper/distr.go:45]: ahhh yessss

Comment (hieuvubk) [reply x/pool-incentives/keeper/distr.go]: Thank you!

File: x/pool-incentives/keeper/distr_test.go
Comment (czarcas7ic) [inline x/pool-incentives/keeper/distr_test.go:90]: The math here wasn't obvious to me until I spun up a local testnet and saw that the community pool param is set to 10%, the staking param is set to 40% (can we change this to say staking instead of fee?), and the gauge param (pool incentives) was set to 30%, with the remaining 20 percent going to dev rewards. Can we make this clearer either through comments or by making this math explicit in the test itself?

File: x/pool-incentives/keeper/distr_test.go
Comment (czarcas7ic) [inline x/pool-incentives/keeper/distr_test.go:123]: same as above comment

**Review by czarcas7ic — CHANGES_REQUESTED** [review]

**Comment by czarcas7ic** [comment]
Good work on making these test cases table driven! Much cleaner! Just a couple of comments to further increase clarity 😄

File: x/pool-incentives/keeper/distr_test.go
Comment (mattverse) [inline x/pool-incentives/keeper/distr_test.go:350]: nit: do you want to try adding test cases for having different weights to check if the total weight is getting updated correctly?

Comment (mattverse) [reply x/pool-incentives/keeper/distr_test.go:350]: And then we can have expectedTotalWeight in the test case field

**Review by mattverse — APPROVED** [review]
Nice work on these TDD! I love it! 🌮 

Left some comments, please take a look!

File: x/pool-incentives/keeper/distr_test.go
Comment (alexanderbez) [reply x/pool-incentives/keeper/distr_test.go:61]: good idea

File: x/pool-incentives/keeper/distr_test.go
Comment (mattverse) [reply x/pool-incentives/keeper/distr_test.go:118]: sweet! There was a remaining test above, could we remove that?

Comment (hieuvubk) [reply x/pool-incentives/keeper/distr_test.go:101]: I already removed the old test and add some comments for more clarity. Thank you for all suggestion.
cc @czarcas7ic @mattverse

File: x/pool-incentives/keeper/distr_test.go
Comment (mattverse) [inline x/pool-incentives/keeper/distr_test.go:66]: nice work on the comments here

**Review by mattverse — APPROVED** [review]
LGTM! This was one of the test cases that desparately needed tdd imo
Nice work!

**Review by alexanderbez — APPROVED** [review]

**Review by czarcas7ic — APPROVED** [review]
My concerns have been addressed, great work @hieuvubk 😄

File: x/pool-incentives/keeper/distr_test.go
Comment (p0mvn) [inline x/pool-incentives/keeper/distr_test.go:143]: What this test seems to be doing is testing `AllocateAsset` vicariously through mint's methods. As a result, this becomes more like a functional test.

I would like to suggest testing this without using mint keeper in general and by calling `AllocateAsset` directly

We can do so by sending some coins to the pool-incentives' module account balance in the test setup.

The fact that the balance of the pool-incentives keeper is updated upon calling `DistributeMintedCoin` is already tested in `x/mint` so there is no need to duplicate that here.

Please let me know what you think

File: x/pool-incentives/keeper/distr_test.go
Comment (p0mvn) [inline x/pool-incentives/keeper/distr_test.go:58]:
```suggestion
				sdk.NewCoins(sdk.NewCoin(sdk.DefaultBondDenom, sdk.NewInt(2500))),
```

**Review by p0mvn — CHANGES_REQUESTED** [review]
Great work @hieuvubk, everything is super clear and easy to follow.

I have some comments. Sorry for blocking this PR but I think it is important that we test `AllocateAsset` directly, not via mint keeper methods. Please let me know what you think

File: x/pool-incentives/keeper/distr_test.go
Comment (hieuvubk) [reply x/pool-incentives/keeper/distr_test.go:143]: I just kept the old logic flow and put the test cases into a table. 
Thanks for letting me know that, it will make the test much easy to look.
Do we want to skip the hook part and call the AllocateAsset function directly?

**Review by p0mvn — APPROVED** [review]
Awesome! Great work

File: x/pool-incentives/keeper/distr_test.go
Comment (p0mvn) [reply x/pool-incentives/keeper/distr_test.go]: Sorry for the late reply. Yes, that's exactly what I meant. What we have right now looks great

**Comment by p0mvn** [comment]
hey @hieuvubk  could you please resolve the conflicts? We can merge this right after

**Comment by hieuvubk** [comment]
> hey @hieuvubk could you please resolve the conflicts? We can merge this right after

yes sure

**Comment by czarcas7ic** [comment]
Good to go, thanks @hieuvubk!

**Pull request merged**
